bindings/cpp/src/ffi_converter.hpp

```diff
@@ -0,0 +1,255 @@
+    ffi_datum.datum_type = static_cast<int32_t>(datum.type);
+    ffi_datum.bool_val = datum.bool_val;
+    ffi_datum.i32_val = datum.i32_val;
+    ffi_datum.i64_val = datum.i64_val;
+    ffi_datum.f32_val = datum.f32_val;
+    ffi_datum.f64_val = datum.f64_val;
+    ffi_datum.string_val = rust::String(datum.string_val);
+
+    rust::Vec<uint8_t> bytes;
+    for (auto b : datum.bytes_val) {
+        bytes.push_back(b);
+    }
+    ffi_datum.bytes_val = std::move(bytes);
+
+    return ffi_datum;
+}
+
+inline ffi::FfiGenericRow to_ffi_generic_row(const GenericRow& row) {
+    ffi::FfiGenericRow ffi_row;
+
+    rust::Vec<ffi::FfiDatum> fields;
+    for (const auto& field : row.fields) {
+        fields.push_back(to_ffi_datum(field));
+    }
+    ffi_row.fields = std::move(fields);
+
+    return ffi_row;
+}
+
+inline Column from_ffi_column(const ffi::FfiColumn& ffi_col) {
+    return Column{
+        std::string(ffi_col.name),
+        static_cast<DataType>(ffi_col.data_type),
+        std::string(ffi_col.comment)};
+}
+
+inline Schema from_ffi_schema(const ffi::FfiSchema& ffi_schema) {
+    Schema schema;
+
+    for (const auto& col : ffi_schema.columns) {
+        schema.columns.push_back(from_ffi_column(col));
+    }
+
+    for (const auto& pk : ffi_schema.primary_keys) {
+        schema.primary_keys.push_back(std::string(pk));
+    }
+
+    return schema;
+}
+
+inline TableInfo from_ffi_table_info(const ffi::FfiTableInfo& ffi_info) {
+    TableInfo info;
+
+    info.table_id = ffi_info.table_id;
+    info.schema_id = ffi_info.schema_id;
+    info.table_path = TablePath{
+        std::string(ffi_info.table_path.database_name),
+        std::string(ffi_info.table_path.table_name)};
+    info.created_time = ffi_info.created_time;
+    info.modified_time = ffi_info.modified_time;
+
+    for (const auto& pk : ffi_info.primary_keys) {
+        info.primary_keys.push_back(std::string(pk));
+    }
+
+    for (const auto& bk : ffi_info.bucket_keys) {
+        info.bucket_keys.push_back(std::string(bk));
+    }
+
+    for (const auto& pk : ffi_info.partition_keys) {
+        info.partition_keys.push_back(std::string(pk));
+    }
+
+    info.num_buckets = ffi_info.num_buckets;
+    info.has_primary_key = ffi_info.has_primary_key;
+    info.is_partitioned = ffi_info.is_partitioned;
+
+    for (const auto& prop : ffi_info.properties) {
+        info.properties[std::string(prop.key)] = std::string(prop.value);
+    }
+
+    info.comment = std::string(ffi_info.comment);
+    info.schema = from_ffi_schema(ffi_info.schema);
+
+    return info;
+}
+
+inline Datum from_ffi_datum(const ffi::FfiDatum& ffi_datum) {
+    Datum datum;
+    datum.type = static_cast<DatumType>(ffi_datum.datum_type);
+    datum.bool_val = ffi_datum.bool_val;
+    datum.i32_val = ffi_datum.i32_val;
+    datum.i64_val = ffi_datum.i64_val;
+    datum.f32_val = ffi_datum.f32_val;
+    datum.f64_val = ffi_datum.f64_val;
+    datum.string_val = std::string(ffi_datum.string_val);
```

Comment:
seem here need to a string copy?

Reply:
I'm not sure is there any effient way. If complex, maybe left a todo to mark it. It'll reminds us if we find any bottle neck